### DISCUSSION TITLE
How to mock different scenarios with Detox?

### DISCUSSION THREAD

**Discussion by lukebrandonfarrell** [wix/Detox#2935] [Q&A] [answered]
A question on how I would mock different scenarios, say I wanted to mock an image picker:

- Mock selection of a PNG
- Mock selection of a JPG
- Mock selection of multiple PNG images

I can write a mock of my image picker which returns a PNG file path in my detox tests, but how would I configure variable behaviour per test run? i.e. say I wanted to test all the three scenarios above with the same mock file. (or different mock files). It seems as if this is not supported.

If this isn't supported, what would be the road to supporting something like this? Are other people interested in this feature? It would need to be controlled with environment variables, as the mock code is setup at compile time, not runtime.

**Comment by d4vidi** [comment, marked as answer]
Hey @lukebrandonfarrell, that's a very interesting topic.
Mocking is not something Detox is suppose to _support_ per se (no Detox feature is associated with that, explicitly). However, it is something so fundamental in testing, that we care about helping out with it as if was a core feature.

In general, the recommended mocking technique is described in our [mocking guide](https://github.com/wix/Detox/blob/master/docs/Guide.Mocking.md#usage). As for the use case you're describing, it sounds to me that it aligns with the technique pretty well. Largely, I think these principles sum it up:

1. Assuming your main React component for image picking is implemented in a file called `ImagePicker.jsx` -- create a mock equivalent in a file called `ImagePicker.e2e.jsx`.
2. To me it sounds like the mocked component should in fact be a screen that contains 3 buttons (for example: `select PNG`, `select JPG`, `select multiple PNGs`). Each button should be assigned a unique test ID.
3. When executing the tests, set RN_SRC_EXT to `e2e.jsx,e2e.js`.
4. In your detox test implementing the selection flow, integrate the mocking screen's behaviour - namely, tap on the proper button associated with each case.

Does that roughly solve you case?

**Reply by lukebrandonfarrell** [reply]
@d4vidi Gotcha, that makes a lot of sense, I wasn't thinking about it in terms of screens, that would solve the problem!

**Reply by d4vidi** [reply]
Good to hear. You're right - the common use case is for working with mock servers, and it's hard to miss that more potential is there 😄

**Reply by lukebrandonfarrell** [reply]
@d4vidi I'm still wondering is there any way to pass an environment parameter to the Mock tests? I understand that the Mocks are compiled at build time, and then the detox JS is executed with node at runtime, but I am wondering could I setup a ENV variable in console and then the Mock would look for that env variable?

**Reply by d4vidi** [reply]
I'm not entirely certain I understood correctly, but here's my answer nonetheless:
1. If you want env vars to be passed into the context of your mock server, then - assuming you're launching it via JS, there are ways to inject env vars onto the child process.
2. If you wish to pass data into the _app_ itself (running on the device), then Detox allows for [passing in custom arguments](https://github.com/wix/Detox/blob/master/docs/APIRef.LaunchArgs.md) into the app.

**Reply by lukebrandonfarrell** [reply]
Thanks @d4vidi so using `device.appLaunchArgs` we can pass arguments from the mock server into the app itself, would that be right?

It seems as if to access these in Android, we'd need to use `currentActivity.getIntent().getBundleExtra("launchArgs").` so to access these in RN, we'd need to write a small custom module? Would I be on the right track here?

For iOS the docs say `On iOS, the specified launch arguments are passed as the process launch arguments and available through normal means`, what would be the normal process (Objective-C) for obtaining launch arguments on iOS? (Maybe we can improve the docs here)

Thanks for drilling down into the details with me !

**Reply by d4vidi** [reply]
That's basically it. If you're using `react-native-navigation` in your project, then you have these available out-of-the-box using `Navigation.getLaunchArgs()`. Otherwise, you can use these references:
- Android: https://github.com/wix/react-native-navigation/blob/master/lib/android/app/src/main/java/com/reactnativenavigation/utils/LaunchArgsParser.java#L19
- iOS: https://github.com/wix/react-native-navigation/blob/master/lib/ios/RNNBridgeModule.m#L231

**Reply by lukebrandonfarrell** [reply]
Thanks @d4vidi !

**Reply by d4vidi** [reply]
Courtesy of @ruben00james: here's a good blogpost about mocking such flows with Detox: https://medium.com/geekculture/how-to-mock-react-native-image-picker-using-detox-9f7ced41b67f

**Reply by pke** [reply]
as for env vars you could use react-native-config and .env files and use those in your react files.

**Comment by ruben00james** [comment]
I have a question:

Can i put the ImagePickerProvider.ts and ImagePickerProvider.e2e.ts files used for mocking in another directory like e2e/mocks? If yes, what do i then need to change to my configuration?

i would like to know this because putting all these files in the root of our project is not ideal.
